### ISSUE TITLE
Installation CentOS 7 failing

### ISSUE THREAD

**Issue by gerners** [biocore/sortmerna#180]
Hi,

I am also trying to install sortmerna, similar to #179 but on a CentOS 7 machine and with a different error. With the new install instructions (thanks!), I progressed much further than last time I tried a few days ago.

However, I also fail after step 6) with libz, rocksdb and rapidjson build from source (within the sortmerna directory).
Command was the following with $SMR_HOME standing for the directory the git release of v3.0.3 was unpacked to as well as where libz, rocksdb and rapidjson have been build.

```
cmake -G "Unix Makefiles" \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_INSTALL_PREFIX=../../dist -DCPACK_BINARY_TGZ=ON \
-DZLIB_ROOT=${SMR_HOME}/zlib/dist \
-DZLIB_LIBRARY_RELEASE=${SMR_HOME}/zlib/dist/lib/libz.a \
-DZLIB_LIBRARY_DEBUG=${SMR_HOME}/zlib/dist/lib/libzd.a \
-DROCKSDB_HOME=${SMR_HOME}/rocksdb/dist \
-DROCKSDB_SRC=${SMR_HOME}/rocksdb \
-DRAPIDJSON_HOME=${SMR_HOME}/rapidjson/dist ../..
```

After performing the first build command I fail with the following error:

```
cmake --build .
[  1%] Built target build_version
[ 37%] Built target cmph
[ 60%] Built target alp
[ 64%] Built target indexdb
[ 96%] Built target smr_objs
[ 98%] Linking CXX executable sortmerna
CMakeFiles/smr_objs.dir/cmd.cpp.o: In function `CmdSession::cmdRead(Runopts&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
cmd.cpp:(.text+0x2440): undefined reference to `rocksdb::DB::Open(rocksdb::Options const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rocksdb::DB**)'
CMakeFiles/smr_objs.dir/cmd.cpp.o: In function `CmdSession::cmdIndex(Runopts&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
cmd.cpp:(.text+0x4027): undefined reference to `rocksdb::DB::Open(rocksdb::Options const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rocksdb::DB**)'
CMakeFiles/smr_objs.dir/output.cpp.o: In function `generateReports(Runopts&, Readstats&, Output&)':
output.cpp:(.text+0x4ad2): undefined reference to `rocksdb::DB::Open(rocksdb::Options const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rocksdb::DB**)'
CMakeFiles/smr_objs.dir/paralleltraversal.cpp.o: In function `align(Runopts&, Readstats&, Output&)':
paralleltraversal.cpp:(.text+0x9a4): undefined reference to `rocksdb::DB::Open(rocksdb::Options const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rocksdb::DB**)'
CMakeFiles/smr_objs.dir/processor.cpp.o: In function `postProcess(Runopts&, Readstats&, Output&)':
processor.cpp:(.text+0x27fa): undefined reference to `rocksdb::DB::Open(rocksdb::Options const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rocksdb::DB**)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [src/sortmerna/sortmerna] Error 1
gmake[1]: *** [src/sortmerna/CMakeFiles/sortmerna.dir/all] Error 2
gmake: *** [all] Error 2
```

Two discrepancies to the manual were observed
1) libz compiled fine, but I do not have a debug libzd.a file as it seems to be required for the first cmake command. The lib folder looks like this:

```
ll ${SMR_HOME}/zlib/dist/lib
total 268K
-rw-r--r--. 1 gerner login 144K Jan 17 13:37 libz.a
lrwxrwxrwx. 1 gerner login    9 Jan 17 13:37 libz.so -> libz.so.1
lrwxrwxrwx. 1 gerner login   14 Jan 17 13:37 libz.so.1 -> libz.so.1.2.11
-rwxr-xr-x. 1 gerner login 115K Jan 17 13:37 libz.so.1.2.11
```
2) The two checks for rocksdb you describe after compiling don't fully match, i.e. after building and popd back to the rocksdb source directory, which I assume the first 'list' command checks, I do not have lib but lib64 such as:
```
ll ${SMR_HOME}/rocksdb/dist/lib64/
total 26M
drwxr-xr-x. 1 gerner login   14 Jan 17 14:05 cmake
-rw-r--r--. 1 gerner login  18M Jan 17 13:55 librocksdb.a
lrwxrwxrwx. 1 gerner login   15 Jan 17 14:05 librocksdb.so -> librocksdb.so.5
lrwxrwxrwx. 1 gerner login   20 Jan 17 14:05 librocksdb.so.5 -> librocksdb.so.5.17.2
-rwxr-xr-x. 1 gerner login 7.3M Jan 17 14:04 librocksdb.so.5.17.2
```

which does have all the libraries, but version 5.17.2 as was checked out according to the manual, however in your example it seems version 5.18.0 was used as listed in the manual. 5.18.0 however is not available as a git tag.

Can you determine where the error does come from? Could it be the wrong rocksdb version number checked out (5.17.2 instead of 5.18.0), leading to undefined reference when building?
Looking forward to the CentOS executable, maybe this one works also on my system which you mentioned to build in #179 .

Thanks a lot in any case and best,
Samuel

**Comment by biocodz** [comment]
I'm looking into this right now and also seeing errors. Seems CMake manages to find something I don't want it to find, and misses intended things.
I had to provide two more flags:
```
 -DZLIB_ROOT_DIR=~/zlib/dist -DZLIB_LIBRARIES=~/zlib/dist/lib/libz.a
```
Now the build is moving on. I'll post the results when done...

**Comment by biocodz** [comment]
Meanwhile,
1.  `5.18.0` is the version on the HEAD [HERE: version.h](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/version.h) i.e. not a release. I'll fix the readme
2. `-DZLIB_LIBRARY_DEBUG` is not really necessary because we are building a Release, and the instructions for ZLIB building also given for release. I'll fix the readme to prevent this confusion.

**Comment by adityabandla** [comment]
On another note, it would be great if you can update the conda recipe for the latest release

**Comment by biocodz** [comment]
@adityabandla 
That's outside the subject of this thread. I'll keep it in mind  but, please, create another issue.

**Comment by biocodz** [comment]
Still working on this. The last error I see:
```
[100%] Linking CXX executable sortmerna
/home/biocodz/rocksdb/dist/lib64/librocksdb.a(env_posix.cc.o): In function `rocksdb::(anonymous namespace)::PosixEnv::NowNanos()':
env_posix.cc:(.text+0x512): undefined reference to `clock_gettime'
collect2: error: ld returned 1 exit status
gmake[2]: *** [src/sortmerna/sortmerna] Error 1
gmake[1]: *** [src/sortmerna/CMakeFiles/sortmerna.dir/all] Error 2
gmake: *** [all] Error 2
```
Such a different in behaviour of CMake on two LInux systems. Even Windows works fine:-)
Keep digging ...

**Comment by biocodz** [comment]
BTW, for rocksdb I had to use 3 flags:
```
-DZLIB_ROOT_DIR=~/zlib/dist \
-DZLIB_LIBRARIES=~/zlib/dist/lib/libz.a \
-DZLIB_LIBRARIES=~/zlib/dist/lib/libz.so \
...
```

**Comment by adityabandla** [comment]
I got this error earlier today. You need to supply -DEXTRA_CXX_FLAGS_RELEASE="-lrt" to CMAKE while building sortmerna. Got this info from your instructions page

**Comment by adityabandla** [comment]
> Still working on this. The last error I see:
> 
> ```
> [100%] Linking CXX executable sortmerna
> /home/biocodz/rocksdb/dist/lib64/librocksdb.a(env_posix.cc.o): In function `rocksdb::(anonymous namespace)::PosixEnv::NowNanos()':
> env_posix.cc:(.text+0x512): undefined reference to `clock_gettime'
> collect2: error: ld returned 1 exit status
> gmake[2]: *** [src/sortmerna/sortmerna] Error 1
> gmake[1]: *** [src/sortmerna/CMakeFiles/sortmerna.dir/all] Error 2
> gmake: *** [all] Error 2
> ```
> Such a difference in behaviour of CMake on two LInux systems. Even Windows works fine:-)
> Keep digging ...

I got this error earlier today. You need to supply -DEXTRA_CXX_FLAGS_RELEASE="-lrt" to CMAKE while building sortmerna. Got this info from your instructions page

**Comment by biocodz** [comment]
Good catch. Forgetting the rocks I hit before:-)

**Comment by biocodz** [comment]
Using  EXTRA_CXX_FLAGS_RELEAS flas is a sub-optimal solution. I'll fix the CMakeList to handle it.

**Comment by biocodz** [comment]
I added the installation files for Centos to the releases.
To sum-up:
Build RocksDB using additional flags:
```
-DZLIB_ROOT_DIR=~/zlib/dist \
-DZLIB_LIBRARIES=~/zlib/dist/lib/libz.a \
-DZLIB_LIBRARIES=~/zlib/dist/lib/libz.so \
...
```
For Sortmerna use additional flag:
```
-DEXTRA_CXX_FLAGS_RELEASE="-lrt" 
```

**Comment by biocodz** [comment]
Need more fixes. I see that ZLIB was actually dynamically linked i.e. not what was intended
```
 ldd ../../dist/bin/sortmerna
        linux-vdso.so.1 =>  (0x00007fffff9ff000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f74196b0000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f7419493000)
        libz.so.1 => /lib64/libz.so.1 (0x00007f741927c000)                             # <-- HERE
        libstdc++.so.6 => /usr/local/lib64/libstdc++.so.6 (0x00007f7418ee5000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f7418c61000)
        libgcc_s.so.1 => /usr/local/lib64/libgcc_s.so.1 (0x00007f7418a4a000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f74186b6000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f74198c3000)
```

**Comment by adityabandla** [comment]
Was it supposed to be a static version?

**Comment by biocodz** [comment]
yep

**Comment by biocodz** [comment]
A note on the `thirdparty.inc`. This file is only intended for use on Windows. 
It will be reflected in the updated Readme.

**Comment by gerners** [comment]
Thank you very much, sorry for the delay.

A colleague managed to get it installed, I personally still did fail, I assume I had some incorrect flags not properly adapted to my system.

Best,
Samuel

**Comment by biocodz** [comment]
@gerners 
If you have problem installing, please, provide the details.
Cheers
biocodz
